src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java

```diff
@@ -366,6 +369,92 @@ String readErrorMessage(InputStream in) throws IOException {
+     * If we get IOE during previous command execution, delay throwing it until
+     * completion status have been read.
+     */
+    void processCompletionStatus(IOException ioe, String cmd, InputStream sis) throws AgentLoadException, IOException {
+        // Read the command completion status
+        int completionStatus;
+        try {
+            completionStatus = readInt(sis);
+        } catch (IOException x) {
+            sis.close();
+            if (ioe != null) {
+                throw ioe;
+            } else {
+                throw x;
+            }
+        }
+        if (completionStatus != 0) {
+            // read from the stream and use that as the error message
+            String message = readErrorMessage(sis);
+            sis.close();
+
+            // In the event of a protocol mismatch then the target VM
+            // returns a known error so that we can throw a reasonable
+            // error.
+            if (completionStatus == ATTACH_ERROR_BADVERSION) {
+                throw new IOException("Protocol mismatch with target VM");
+            }
+
+            // Special-case the "load" command so that the right exception is
+            // thrown.
+            if (cmd.equals("load")) {
+                String msg = "Failed to load agent library";
+                if (!message.isEmpty()) {
+                    msg += ": " + message;
+                }
+                throw new AgentLoadException(msg);
+            } else {
+                if (message.isEmpty()) {
+                    message = "Command failed in target VM";
+                }
+                throw new AttachOperationFailedException(message);
+            }
+        }
+    }
+
+    /*
+     * InputStream for the socket connection to get target VM
+     */
+    abstract static class SocketInputStream extends InputStream {
+        private final long fd;
+
+        public SocketInputStream(long fd) {
+            this.fd = fd;
+        }
+
+        protected abstract int read(long fd, byte[] bs, int off, int len) throws IOException;
+        protected abstract void close(long fd) throws IOException;
+
+        public synchronized int read() throws IOException {
+            byte b[] = new byte[1];
+            int n = this.read(b, 0, 1);
+            if (n == 1) {
+                return b[0] & 0xff;
+            } else {
+                return -1;
+            }
+        }
+
+        public synchronized int read(byte[] bs, int off, int len) throws IOException {
+            if ((off < 0) || (off > bs.length) || (len < 0) ||
+                ((off + len) > bs.length) || ((off + len) < 0)) {
+                throw new IndexOutOfBoundsException();
+            } else if (len == 0) {
+                return 0;
+            }
+            return read(fd, bs, off, len);
+        }
+
+        public synchronized void close() throws IOException {
+            if (fd != -1) {
+                close(fd);
+            }
```

Comment:
There used to be logic to set `fd` (previously called `s`) to -1 during the close operation. Is that no longer needed?

Reply:
I don't see why we need it because this method is protected by synchronized.

Reply:
You will end up leaving fd set even after the close completes.  A subsequent close (which I assume should never happen) would end up with an IOException. Previously close would end up not doing anything when called again, and simply return with no error. I'm not sure which is the best behavior in this case, but it is a change in behavior.